src/librustc_middle/ty/layout.rs

```diff
@@ -877,56 +895,121 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
+                    if struct_reordering_opt {
+                        let mut extend_niche_range = |d| {
+                            niche_variants =
+                                *niche_variants.start().min(&d)..=*niche_variants.end().max(&d);
+                        };
+                        let mut align = dl.aggregate_align;
 
-                    // Find one non-ZST variant.
-                    'variants: for (v, fields) in variants.iter_enumerated() {
-                        if absent(fields) {
-                            continue 'variants;
+                        // Find the largest and second largest variant.
+                        for (v, fields) in variants.iter_enumerated() {
+                            if absent(fields) {
+                                continue;
+                            }
+                            let mut size = Size::ZERO;
+                            for &f in fields {
+                                align = align.max(f.align);
+                                size += f.size;
+                            }
+                            if size > max_size {
+                                second_max_size = max_size;
+                                max_size = size;
+                                if let Some(d) = dataful_variant {
+                                    extend_niche_range(d);
+                                }
+                                dataful_variant = Some(v);
+                            } else if size == max_size {
+                                if let Some(d) = dataful_variant {
+                                    extend_niche_range(d);
+                                }
+                                dataful_variant = None;
+                                extend_niche_range(v);
+                            } else {
+                                second_max_size = second_max_size.max(size);
+                                extend_niche_range(v);
+                            }
                         }
-                        for f in fields {
-                            if !f.is_zst() {
-                                if dataful_variant.is_none() {
-                                    dataful_variant = Some(v);
-                                    continue 'variants;
-                                } else {
-                                    dataful_variant = None;
-                                    break 'variants;
+                        let padding_bits = (max_size.align_to(align.abi) - max_size).bits();
+                        let count =
+                            niche_variants.end().as_u32() - niche_variants.start().as_u32() + 1;
+                        if padding_bits >= 32 || count >> padding_bits == 0 {
+                            // Don't perform niche optimisation if there is enough padding bits for the discrimiment
+                            dataful_variant = None;
+                        }
```

Comment:
So this is a more correct version of the "only use niche if it's *physically impossible* to fit a tag anywhere", right?

It's still conservative in that it uses the tag layout unless *it knows for sure* that the niche layout would be beneficial, right?

Reply:
Yes that's right.
I changed it again to be sure than the padding is actually bigger than the biggest niche so we'd still use the niche in `Option2<NicheU16, u8>`  (which i meant instead of `Option<(NicheU16, u8)>` from my previous comment which is its own type as you said)